sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs

```diff
@@ -218,7 +228,17 @@ protected EventHubProducer()
             Guard.ArgumentNotNull(nameof(events), events);
             GuardSinglePartitionReference(PartitionId, options.PartitionKey);
 
-            return InnerProducer.SendAsync(events, options, cancellationToken);
+            using DiagnosticScope scope = CreateDiagnosticScope();
+            InstrumentMessages(events);
```

Comment:
We should consider doing a `ToList` or some other transform, since we're now enumerating this twice and have no guarantee that it is safe to do so, no?

Reply:
Since we're only really interested in the side effect for the events, maybe it would make sense for a minor abuse of LINQ to avoid having to talk the set twice.   Something like:

```c#
var activity = Activity.Current;
var instrumentedEvents = events.Select(eventData => 
{
      EventInstrumentation.InstrumentEvent(eventData, activity);
      return eventData;
});

//...
await InnerProducer.SendAsync(instrumentedEvents , options, cancellationToken);
// ...
```

Reply:
I don't think mutating as part of enumeration is appropriate here as we don't know what happens with enumerable downrange. The track 1 client enumerates events multiple times...

Reply:
The track one client is a bad model.   We're accepting an unknown enumeration in a public API that may or may not be safe to walk multiple times.   Generally, my understanding is that unless you're in control of the type and know for sure that it is safe, you want to either do a straight transform to a type that is safe or use the deferred translation to run during enumeration to avoid multiples.   This is, of course, abusing the deferred map for a side effect, but it would potentially save doing 3 walks through the set.   That said, we know the set is not of a large size, so you could make the case that just doing a `ToList` and then using what we have is probably as good, if not potentially better, than adding the LINQ into it.

Reply:
I've added `ToList()`, we can rethink if it becomes a problem.